### DISCUSSION TITLE
Exploit common computations in objective and constraints in `nlpsol`?

### DISCUSSION THREAD

**Discussion by gregreich** [casadi/casadi#3627] [General]
Hi

Is there any way to exploit the presence of computations that are common to the objective function and the constraints in the `nlpsol` class (in Matlab)?

Some solver interfaces (outside of the Casadi universe) allow to specify objective function and constraints as one function with multiple outputs; but I couldn't figure how to do it with Casadi.

Thanks!

**Comment by jaeandersson** [comment]
There is nothing in the CasADi `nlpsol` class that implies that objectives and constraints cannot have common subexpressions and thus common computations. For example, if you calculate the Hessian of the Lagrangian in the IPOPT interface, objective and constraints are calculated together. It's a feature of the specific solvers (like IPOPT) whether they want to calculate `f`, `g` or both of them together.

As a side note, I doubt that you get much efficiency improvements by calculating `f` and `g` together since for most application the bulk of the computational time (at least with IPOPT) will be in (1) IPOPT's linear solver, (2) Jacobian of `g` and (3) Hessian of L = f + dot(lam_g, g). It is definitely possible to get significant efficiency improvements by calculating Jacobian and Hessian together, but that would require a some work on the CasADi side to really exploit.

**Reply by gregreich** [reply]
Thank you, @jaeandersson, for your prompt reply!

> It is definitely possible to get significant efficiency improvements by calculating Jacobian and Hessian together, but that would require a some work on the CasADi side to really exploit.

If this is the bottleneck, I assume introducing some auxiliary variable, say `z`, set it equal to the common expressions in `f` and `g` via a set of non-linear constraints, and then only use `z` within `f` and `g` (instead of the expressions themselves) should allow for this to be exploited, right? (at the expensive of a potentially more difficult problem, of course)

**Reply by jaeandersson** [reply]
No. You can get a common calculation of f and g in ipopt this way (with a dummy objective), noting today ipopt might not treat both problems equivalently. Common calculation of J and H is a separate problem, which you have even if the objective is trivial.

**Reply by gregreich** [reply]
OK, understood. 

Thanks again, Joel!
